sway-lsp/src/core/sync.rs

```diff
@@ -23,88 +22,101 @@ pub enum Directory {
     Temp,
 }
 
-#[derive(Debug)]
+#[derive(Debug, Default)]
 pub struct SyncWorkspace {
     pub directories: DashMap<Directory, PathBuf>,
 }
 
 impl SyncWorkspace {
     pub const LSP_TEMP_PREFIX: &'static str = "SWAY_LSP_TEMP_DIR";
 
-    pub(crate) fn new() -> Self {
-        Self {
-            directories: DashMap::new(),
-        }
+    pub fn new() -> Self {
+        Self::default()
     }
 
-    /// Overwrite the contents of the tmp/folder with everything in
-    /// the current workspace.
-    pub fn resync(&self) -> Result<(), LanguageServerError> {
-        self.clone_manifest_dir_to_temp()?;
-        self.sync_manifest();
-        Ok(())
-    }
-
-    /// Clean up the temp directory that was created once the
-    /// server closes down.
-    pub(crate) fn remove_temp_dir(&self) {
+    /// Clean up the temp directory that was created once the server closes down.
+    pub fn remove_temp_dir(&self) {
         if let Ok(dir) = self.temp_dir() {
-            dir.parent().map(fs::remove_dir);
+            // The `temp_path` we store is `random_dir/project_name`.
+            // So, we need to remove `random_dir` by getting the parent directory.
+            if let Some(parent_dir) = dir.parent() {
+                if parent_dir
+                    .to_string_lossy()
+                    .contains(SyncWorkspace::LSP_TEMP_PREFIX)
```

Comment:
This string-based check may produce false positives/negatives if another directory in the path happens to contain the prefix as a substring.
Prefer testing only the directory’s own name.
i.e.
```suggestion
                if parent_dir
                    .file_name()
                    .is_some_and(|name| name.to_string_lossy().starts_with(SyncWorkspace::LSP_TEMP_PREFIX))
```
This is safer and avoids accidental deletion of unrelated directories.